org.sonarlint.eclipse.ui/src/org/sonarlint/eclipse/ui/internal/backend/SonarLintEclipseClient.java

```diff
@@ -161,20 +165,36 @@ public void showHotspot(ShowHotspotParams params) {
   @Override
   public CompletableFuture<AssistCreatingConnectionResponse> assistCreatingConnection(AssistCreatingConnectionParams params) {
     return CompletableFuture.supplyAsync(() -> {
+      var baseUrl = params.getServerUrl();
+      
       try {
+        AbstractAssistCreatingConnectionJob job;
+        
         SonarLintLogger.get().debug("Assist creating a new connection...");
-        var job = new AssistCreatingConnectionJob(params.getServerUrl());
+        if (params.getTokenName() != null && params.getTokenValue() != null) {
+          job = new AssistCreatingAutomaticConnectionJob(baseUrl, params.getTokenValue());
+        } else {
+          job = new AssistCreatingManualConnectionJob(baseUrl);
+        }
+        
         job.schedule();
         job.join();
         if (job.getResult().isOK()) {
           SonarLintLogger.get().debug("Successfully created connection '" + job.getConnectionId() + "'");
           return new AssistCreatingConnectionResponse(job.getConnectionId());
-        } else if (job.getResult().getCode() == IStatus.CANCEL) {
+        } else if (job.getResult().matches(IStatus.CANCEL)) {
+          if (job instanceof AssistCreatingAutomaticConnectionJob) {
+            SonarLintBackendService.get()
+              .getBackend()
+              .getUserTokenService()
+              .revokeToken(new RevokeTokenParams(baseUrl, params.getTokenName(), params.getTokenValue()));
```

Comment:
This is a notification, right? If not, you should call `join()` on the CompletableFuture.

Reply:
Yes, it is. There is no return value and in the end, the likelihood that it fails is near then zero so it shouldn't block the thread but it can take some time due to the network connection.

It can only _fail_ on three occasions:
- network down (unlikely as we just had contact from the SQ to invoke the "Open in IDE", maybe someone let the confirmation window open and turned off the network connection - highly unlikely)
- the server is down (same as above)
- unauthorized (token was just granted, someone on SQ must have removed it just seconds ago)

Reply:
If it is a notification, it will be the responsibility of the core to handle the errors, so that's fine.